stdio/file.c

```diff
@@ -416,67 +513,167 @@ size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream)
 }
 
 
+static inline size_t buffer_data(FILE *stream, const void *ptr, size_t writesz)
+{
+	size_t bytes = min(stream->bufsz - stream->bufpos, writesz);
+	if (bytes > 0) {
+		memcpy(stream->buffer + stream->bufpos, ptr, bytes);
+		stream->bufpos += bytes;
+	}
+	return bytes;
+}
+
+
+static inline ssize_t write_buffer(FILE *stream, size_t writesz)
+{
+	ssize_t err = full_write(stream->fd, stream->buffer, writesz);
+	if (err >= 0) {
+		stream->bufpos -= err;
+
+		if (err < writesz) {
+			/* EAGAIN */
+			stream->flags |= F_ERROR;
+
+			if (err > 0) {
+				memmove(stream->buffer, stream->buffer + err, stream->bufpos);
+			}
+		}
+	}
+	else {
+		stream->flags |= F_ERROR;
+	}
+	return err;
+}
+
+
+static inline ssize_t write_data(FILE *stream, const void *ptr, size_t writesz)
+{
+	ssize_t err = full_write(stream->fd, ptr, writesz);
+	if (err >= 0) {
+		if (err < writesz) {
+			/* EAGAIN */
+			stream->flags |= F_ERROR;
+		}
+	}
+	else {
+		stream->flags |= F_ERROR;
+	}
+	return err;
+}
+
+
 size_t fwrite_unlocked(const void *ptr, size_t size, size_t nmemb, FILE *stream)
 {
-	int err;
+	ssize_t err;
 	size_t writesz = nmemb * size;
+	size_t total = 0;
+	size_t bytes;
 	char *nl;
 
-	if (!writesz) {
+	if (writesz == 0) {
 		return 0;
 	}
 
 	if (stream->buffer == NULL) {
 		/* unbuffered write */
-		if ((err = __safe_write(stream->fd, ptr, writesz)) < 0) {
-			return 0;
-		}
+		err = write_data(stream, ptr, writesz);
+		return (err > 0) ? err / size : 0;
+	}
 
-		return err / size;
+	if ((stream->mode & O_RDONLY) != 0) {
+		set_errno(-EBADF);
+		stream->flags |= F_ERROR;
+		return 0;
 	}
 
-	/* discard reading buffer */
-	if (!(stream->flags & F_WRITING)) {
-		fflush_unlocked(stream);
+	/* flush the read buffer if currently reading */
+	if ((stream->flags & F_WRITING) == 0) {
+		if (__fflush_one(stream) < 0) {
+			return 0;
+		}
+		if (stream->bufpos != stream->bufeof) {
+			/* abort if the read buffer cannot be flushed (non-seekable stream) */
```

Comment:
Shouldn't `errno` be set then? (And maybe the error flag?)

Reply:
It's implementation-specific behavior. In this case, I'm emulating glibc.

Reply:
Ugh, case similar to `errno` in `malloc(0)`, so we probably can't solve this reasonably. Ack.